Internal java solution for CSES Apartments (#1952)
cpinitiative/usaco-guide: solutions/cses-1084.mdx

id: cses-1084
source: CSES
title: Apartments
author: Nathan Gong, Danh Ta Chi Thanh
---

[Unofficial Editorial](https://codeforces.com/blog/entry/83295)

We can use a greedy approach to solve this problem. First, sort the applicant
and apartment arrays. We can keep two pointers $i$ and $j$ (initialized to 0),
which keep track of the current index of the applicant and apartment we are
looking at respectively. Then, while there are more applicants and apartments to
look through, we repeatedly run the following check:

- If $|\texttt{applicants}[i] - \texttt{apartments}[j]| <= k$, we have found a
  suitable apartment for the current applicant. Thus, we increment $i$, $j$, and
  our answer.
- Otherwise, $\texttt{apartments}[j]$ is either too big or too small for
  $\texttt{applicants}[i]$. We can increment $i$ and $j$ accordingly.

## Implementation

**Time Complexity:** $\mathcal{O}(n\log(n)+m\log(m))$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;

// Constant initialization

> const int maxn = 2e5+10;

Reviewer: ```suggestion
const int maxn = 2e5 + 10;
```

> const int maxn = 2e5 + 10;

Reviewer 2: ```suggestion
const int MAX_N = 2e5
```
think this should still work? the +10 makes readers question "wait why is the +10 there"
it's like adding unnecessary supports, people will assume that w/o them the building would fail and then look to see why they're necessary

// Variables used for the current problem

> int n, m, k, a[maxn], b[maxn], ans;

Reviewer 2: better variable names?

void solve() {
	cin >> n >> m >> k;

> 	for (int i = 0;i < n; ++i) cin >> a[i];

Reviewer: ```suggestion
	for (int i = 0; i < n; ++i) cin >> a[i];
```

	for (int i = 0; i < m; ++i) cin >> b[i];

> 	sort(a, a + n); sort(b, b + m);

Reviewer 2: each statement on own line plz

id: cses-1084
source: CSES
title: Apartments
author: Nathan Gong, Danh Ta Chi Thanh
---

> [Unofficial Editorial](https://codeforces.com/blog/entry/83295)

Reviewer 3: ```suggestion
[Unofficial Editorial (C++)](https://codeforces.com/blog/entry/83295)
```

> We can use a greedy approach to solve this problem. First, sort the applicant
> and apartment arrays. We can keep two pointers $i$ and $j$ (initialized to 0),

Reviewer 4: ```suggestion
We can use a greedy approach to solve this problem. First, sort the applicants
and apartments. We can keep two pointers $i$ and $j$ (initialized to 0),
```

which keep track of the current index of the applicant and apartment we are
looking at respectively. Then, while there are more applicants and apartments to
look through, we repeatedly run the following check:

- If $|\texttt{applicants}[i] - \texttt{apartments}[j]| \leq k$, we have found a
  suitable apartment for the current applicant. Thus, we increment $i$, $j$, and
  our answer.
- Otherwise, $\texttt{apartments}[j]$ is either too big or too small for

>   $\texttt{applicants}[i]$. We can increment $i$ and $j$ accordingly.

Reviewer 4: ```suggestion
  $\texttt{applicants}[i]$. We can increment $i$ or $j$ accordingly.
```

## Implementation

**Time Complexity:** $\mathcal{O}(n\log(n)+m\log(m))$

<LanguageSection>

<CPPSection>

```cpp
#include <bits/stdc++.h>

using namespace std;

> // Constant initialization

Reviewer 2: not necessary?

Reviewer 5: not sure what you mean. howcome

Reviewer 2: the comment- i mean it's pretty clear this is initializing a constant

	return 0;
}
```

</CPPSection>

<JavaSection>

> Note: Java will TLE on test case #17.

Reviewer 6: this would be better in a warning maybe?

```java
import java.io.*;
import java.util.*;

public class Apartments {
	public static void main(String[] args) {
		Kattio io = new Kattio();

		int n = io.nextInt();
		int m = io.nextInt();
		int k = io.nextInt();

		int[] applicants = new int[n];
		int[] apartments = new int[m];
		for (int i = 0; i < n; i++) {
			applicants[i] = io.nextInt();
		}
		for (int i = 0; i < m; i++) {
			apartments[i] = io.nextInt();
		}

		Arrays.sort(applicants);
		Arrays.sort(apartments);

		int i = 0;
		int j = 0;
		int ans = 0;
		while (i < n && j < m) {
			// Found suitable apartment
			if (Math.abs(applicants[i] - apartments[j]) <= k) {
				i++;
				j++;
				ans++;
			}
			// Apartment is too small -> increment apartment pointer
			else if (applicants[i] > apartments[j]) {
				j++;
			}
			// Apartment is too big -> increment applicant pointer
			else {
				i++;
			}
		}

> 

Reviewer 6: ```suggestion
```

Author: I like leaving spacing here to separate variables and methods/static classes

Author: Oops I meant to refer to the other suggestion

		io.println(ans);
		io.close();
	}

> 

Reviewer 6: ```suggestion
```

	return 0;
}
```

</CPPSection>

<JavaSection>

<Warning>
Java will TLE on test case #17.
</Warning>

```java
import java.io.*;
import java.util.*;

public class Apartments {
	public static void main(String[] args) {
		Kattio io = new Kattio();

		int n = io.nextInt();
		int m = io.nextInt();

> 		int k = io.nextInt();

Reviewer 2: better names here too

Reviewer: these are from the problem statement though...

Reviewer 2: don't think anyone wants to constantly flip from the problem statement to the solution code over & over again

Author: If they're trying to solve the problem, won't they already know what all of the variables mean?

Reviewer 2: think most people just look at the problem variable names (like "the number of apartments") and go right to thinking + coding

Reviewer 7: ^

Reviewer 7: woops that was meant to point towards @nathangong

Reviewer 2: lmao

Author: > think most people just look at the problem variable names (like "the number of apartments") and go right to thinking + coding

Well the input format is given in terms of n, m, and k so I think it's safe to assume they know what is what given that they have to deal with them to read in the input. Plus, you generally use the variable names n, m, and k when you're trying to determine what the time complexity of your approach is.

Reviewer 2: still think it'd be better to be explicit about the variable names
`peopleNum`, `aptNum`, and `maxDiff` or smth
also what's `a` and `b`

Author: Alright I can compromise by adding comments for each variable explaining what they do. `a` stores the desired appartment size of each applicant and `b` stores the size of each apartment as explained in the problem.

Reviewer 2: yeah cool i'll accept that
